### PR TITLE
feat(xml): oobleck xml parser updates

### PR DESCRIPTION
**Pull request by kulgan**
Summary
To parse XML:
- read mapping from a yaml file in `gdcdatamodel` (the mapping contains node properties names and the xpath in the xml to find the value).
- read xml string into etree xml object
- use the mapping to retrieve node properties and create a node json

This approach only allowed for a single json for a given node type and assumes property and be read from the XML using a simple xpath query

Changes:
- need to create 2 treatment nodes
- properties needs to be retrieved using some rather complex logic

Approach:
Added extra field to the mapping that allows code to decide at runtime an evaluator to use in evaluating a node property or the node itself, so that custom logic can be implemented easily. Added both field and node evaluators which can be extended with relative ease in the future
Also, removed psqlgraph instance dependency, so it can be run as a standalone script

### CODE REVIEW

File: sheepdog/xml/evaluators/fields.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            val = self.search(self.root_element, xpath)
+
+            if self.is_radiation(xpath):
+                val = self.search_additional_paths(val, self.get_evaluator_property("additional_radiation_path"))
+                treatments[0] = dict(treatment_type="Radiation, NOS", treatment_or_therapy=val)
+            else:
+                val = self.search_additional_paths(val, self.get_evaluator_property("additional_pharmaceutical_path"))
+                treatments[1] = dict(treatment_type="Pharmaceutical Therapy, NOS", treatment_or_therapy=val)
+
+        return treatments
```
Comment (anmaxvl) [inline sheepdog/xml/evaluators/fields.py]: is it ok to have them both set? they are not exclusive to each other or anything?

File: sheepdog/xml/evaluators/__init__.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        self.property_mappings = mappings
+        self.xml_namespaces = namespaces
+
+    @property
+    def path(self):
+        return self.property_mappings.get("path")
+
+    @property
+    def is_nullable(self):
+        return self.property_mappings.get("nullable") in [None, "true", "t", "yes"]
```
[the remark this answers is not in the document]
Comment (kulgan) [reply sheepdog/xml/evaluators/__init__.py]: Completely managed by us, it gets pulled from https://github.com/NCI-GDC/gdcdatamodel/blob/feat/DAT-2076-oobleck-xml-changes/gdcdatamodel/xml_mappings/tcga_clinical.yaml

File: sheepdog/xml/evaluators/__init__.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        # check defaults
+        if val is None and self.default:
+            val = self.default
+
+        # check nullable
+        if val is None and not self.is_nullable:
+            raise Exception("Can't find element {}".format(self.path))
+
+        # check if suffix can be appended
+        if self.suffix and self.data_type == "str":
```
[the remark this answers is not in the document]
Comment (kulgan) [reply sheepdog/xml/evaluators/__init__.py]: data_type is read from the mapping and there is code that tries to map it to a function already. Making it an `enum` may just be extra code :trollface:

File: sheepdog/xml/evaluators/__init__.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    def search_path(self, path=None, nullable=True):
+
+        paths = path or self.path
+        if isinstance(paths, str):
+            return self.root_element.xpath(paths, namespaces=self.xml_namespaces, nullable=nullable)
+        raise ValueError("{} property must be a string, but {} found".format(self.__class__.__name__, paths))
+
+    @staticmethod
+    def _to_bool(val):
+        possible_true_values = ['true', 'yes']
```
[the remark this answers is not in the document]
Comment (kulgan) [reply sheepdog/xml/evaluators/__init__.py]: Makes sense to add those too

File: sheepdog/xml/evaluators/fields.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    def get_max_element(self, path):
+        tie_breaker = "sequence"
+        elements = self.search_path(path)
+
+        if not isinstance(elements, list):
+            raise ValueError("path expression must produce a list and not a single value")
+
+        _max, _max_element = None, None
+        for element in elements:
```
[the remark this answers is not in the document]
Comment (anmaxvl) [reply sheepdog/xml/evaluators/fields.py]: discussed this offline. weird stuff.

File: sheepdog/xml/evaluators/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anmaxvl) [reply sheepdog/xml/evaluators/__init__.py]: yeah, sure.

File: sheepdog/xml/evaluators/nodes.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        return self.get_data()
+
+    def get_data(self):
+        new_data = []
+        for key, data in self.data.items():
+            if not isinstance(data, list):
+                continue
+
+            for i, entry in enumerate(data):
+                treatment_data = copy.deepcopy(self.data)
```
Comment (anmaxvl) [inline sheepdog/xml/evaluators/nodes.py]: is this correct? the `self.data` part?

File: sheepdog/xml/evaluators/fields.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            val = "yes"
+        elif [e for e in elements if e.text is not None and e.text.lower() == "no"]:
+            val = "no"
+        return val
+
+    def _evaluate(self):
+
+        # list type expected
+        xpaths = self.path if isinstance(self.path, list) else [self.path] # type: list[str]
+        treatments = [None, None]
```
[the remark this answers is not in the document]
Comment (kulgan) [reply sheepdog/xml/evaluators/fields.py]: With the current setup, 2 treatment nodes will always get created cos the mapping has 2 hard coded paths (a radiation path and a pharm path). If they are not found in the xml, the will be defaulted to not reported

File: sheepdog/xml/evaluators/nodes.py (hunk 6 above)
Comment (kulgan) [reply sheepdog/xml/evaluators/nodes.py]: self.data holds the complete treament json with the extra key used to specified the fields that would be different for the new treatment data, so the copy is to get the base data, the pop is to remove the unused key and the then update

**Pull request merged**